cladecombiner/aggregator.py

```diff
@@ -0,0 +1,228 @@
+from abc import ABC, abstractmethod
+from collections.abc import Iterable
+from warnings import warn
+
+from .taxon import Taxon
+from .taxon_utils import sort_taxa
+from .taxonomy_scheme import PhylogeneticTaxonomyScheme
+from .utils import table
+
+
+class Aggregation(dict[Taxon, Taxon]):
+    """
+    An object for aggregations, basically just a dictionary.
+    """
+
+    def _validate(
+        self, input_taxa: Iterable[Taxon], taxon_map: dict[Taxon, Taxon]
+    ):
+        """
+        Checks that all input taxa have been mapped exactly once.
+        """
+        if set(taxon_map.keys()) != set(input_taxa):
+            raise RuntimeError(
+                "Mismatch between aggregated taxa and input taxa. Input taxa are: "
+                + str(input_taxa)
+                + " but aggregated taxa are "
+                + str(taxon_map.keys())
+            )
+        tab = table(taxon_map)
+        if not all(v == 1 for v in tab.values()):
+            raise RuntimeError(
+                "Found following taxa mapped more than once: "
+                + str([k for k, v in tab.items() if v > 1])
+            )
+
+    def __init__(
+        self, input_taxa: Iterable[Taxon], taxon_map: dict[Taxon, Taxon]
+    ):
+        self._validate(input_taxa, taxon_map)
+        super().__init__(taxon_map)
+
+    def to_str(self):
+        """
+        Get str : str map of taxa names
+        """
+        return {k.name: v.name for k, v in self.items()}
+
+
+class Aggregator(ABC):
+    """
+    Aggregators return Aggregations, maps of input_taxon : aggregated_taxon
+    """
+
+    @abstractmethod
+    def aggregate(self, input_taxa: Iterable[Taxon]) -> Aggregation:
+        pass
+
+
+class ArbitraryAggregator(Aggregator):
+    """
+    Aggregation via a user-provided dictionary.
+    """
+
+    def __init__(
+        self,
+        map: dict[Taxon, Taxon],
+    ):
+        """
+        FixedAggregator constructor.
+
+        Parameters
+        ----------
+        map : dict[Taxon, Taxon]
+            Dictionary mapping the input taxa to their aggregated taxa.
+        """
+        self.map = map
+
+    def aggregate(self, input_taxa: Iterable[Taxon]) -> Aggregation:
+        return Aggregation(
+            input_taxa, {taxon: self.map[taxon] for taxon in input_taxa}
+        )
+
+
+class BasicPhylogeneticAggregator(Aggregator):
```

Comment:
Personally I would prefer just "PhylogeneticAggregator," because I can imagine some more "basic" forms, as per my comments about composability below

Reply:
The name can change, but we have to be able to clearly distinguish between this implicitly fixed-map "put these in those" phylogenetic aggregation and the (probably multiple different) dynamic "put these in something" aggregator(s) that we want to make. My instinct is to qualify all of them (this one being "basic" or "fixed" or something of the sort), but I could be convinced that only the more complex ones need more explicit names.